### PR TITLE
Implement CupertinoMenu and CupertinoNestedMenu

### LINKED ISSUE #60298: Pull-Down Menus for iOS 14
On iOS 14, Pull-Down Menus have been added.

> Menus offer several advantages over action sheets, context menus, and popovers. For example:
>
> A menu opens very near the button that reveals it, so people can instantly understand the relationship between the menu's items and the action they're performing.
>
> In addition to listing actions, a menu can offer selections people can use to affect the primary action.
>
> Menus animate quickly into view and don't dim the screen when they appear, giving both the transition and the overall experience a lightweight feel.

Image

Image

https://developer.apple.com/design/human-interface-guidelines/pull-down-buttons

### PR DESCRIPTION
**Pull request by davidhicks980**
Resolves https://github.com/flutter/flutter/issues/60298, https://github.com/notDmDrl/pull_down_button/issues/26

This PR adds a CupertinoMenu, CupertinoNestedMenu, and associated widgets that increase feature parity with iOS. 

I'm submitting a draft PR to collect feedback on the overall API and layout design before writing more extensive layout and API tests. I have added a few tests based on the PopupMenu widget, and I've done some manual accessibility testing across iOS, Android, and MacOS. Comments and an example have been added, but could use review. I also added TODO annotations for any particularly pressing questions I have. 

I'm still fairly new to Flutter, so I apologize for any unconventional code. Let me know if there are any pieces we need to tear out or rethink. I'm interested to hear different ideas regarding how to solve this menu, so I appreciate you all taking a look!

Also, I used @notDmDrl's repo as a base for my contributions: https://github.com/notDmDrl/pull_down_button, so I'm not sure whether I should contact him with regards to signing a CLA. 

Basic example: 
```dart
import 'package:flutter/cupertino.dart';

/// Flutter code sample for [CupertinoMenuButton]
void main() => runApp(const CupertinoMenuApp());
class CupertinoMenuApp extends StatelessWidget {
  const CupertinoMenuApp({super.key});
  @override
  Widget build(BuildContext context) {
    return const CupertinoApp(
      theme: CupertinoThemeData(brightness: Brightness.light),
      home: CupertinoMenuExample(),
    );
  }
}

class CupertinoMenuExample extends StatefulWidget {
  const CupertinoMenuExample({super.key});

  @override
  State<CupertinoMenuExample> createState() => _CupertinoMenuExampleState();
}

class _CupertinoMenuExampleState extends State<CupertinoMenuExample> {
  String _checkedValue = 'Cat';
  @override
  Widget build(BuildContext context) {
    return CupertinoPageScaffold(
      navigationBar: const CupertinoNavigationBar(
        middle: Text('CupertinoMenu Sample'),
      ),
      child: Align(
        alignment: Alignment(0, -0.5),
        child: SizedBox(
          width: 100,
          height: 100,
          child:  CupertinoMenuButton<String>(
            itemBuilder: (BuildContext context) {
              return <CupertinoMenuEntry<String>>[
                CupertinoStickyMenuHeader(
                  leading: Container(
                    width: 42,
                    height: 42,
                    decoration:  ShapeDecoration(
                      shape: CircleBorder(),
                      gradient: LinearGradient(
                        begin: Alignment.topCenter,
                        end: Alignment.bottomCenter,
                        colors: <Color>[
                          CupertinoColors.activeBlue,
                          CupertinoColors.activeBlue.withOpacity(0.5),
                        ],
                      ),
                    ),
                  ),
                  trailing: CupertinoButton(
                    minSize: 34,
                    padding: EdgeInsets.zero,
                    child: Container(
                      width: 34,
                      height: 34,
                      decoration: ShapeDecoration(
                        shape: const CircleBorder(),
                        color: CupertinoColors.systemFill.resolveFrom(context),
                      ),
                      child: Icon(
                        CupertinoIcons.share,
                        color: CupertinoColors.label.resolveFrom(context),
                        size: 16,
                        semanticLabel: 'Triangle',
                      ),
                    ),
                    onPressed: () {},
                  ),
                  subtitle: const Text('Folder'),
                  child: const Text('Downloads'),
                ),
                CupertinoNestedMenu<String>(
                  itemBuilder: (BuildContext context) => <CupertinoMenuEntry<String>>[
                    CupertinoCheckedMenuItem<String>(
                      value: 'Cat',
                      checked: _checkedValue == 'Cat',
                      shouldPopMenuOnPressed: false,
                      onTap: (){
                        setState(() {
                          _checkedValue = 'Cat';
                        });
                      },
                      child: const Text('Cat'),
                    ),
                    CupertinoCheckedMenuItem<String>(
                      value: 'Feline',
                      checked: _checkedValue == 'Feline',
                      onTap: (){
                        setState(() {
                          _checkedValue = 'Feline';
                        });
                      },
                      shouldPopMenuOnPressed: false,
                      child: const Text('Feline'),
                    ),
                    CupertinoCheckedMenuItem<String>(
                      value: 'Kitten',
                      checked: _checkedValue == 'Kitten',
                      onTap: (){
                        setState(() {
                          _checkedValue = 'Kitten';
                        });
                      },
                      shouldPopMenuOnPressed: false,
                      child: const Text('Kitten'),
                    ),
                  ],
                  subtitle:  Text(_checkedValue),
                  title: const TextSpan(text: 'Favorite Animal'),
                ),
                CupertinoMenuItem<String>(
                  trailing: const Icon(
                    CupertinoIcons.textformat_size,
                  ),
                  child: const Text(
                    'Simple',
                  ),
                  onTap: () {},
                ),
                const CupertinoMenuItem<String>(
                  shouldPopMenuOnPressed: false,
                  trailing: Icon(
                    CupertinoIcons.textformat_size,
                  ),
                  isDefaultAction: true,
                  child: Text('Default'),
                ),
                const CupertinoMenuItem<String>(
                  trailing: Icon(CupertinoIcons.cloud_upload),
                  value: 'Disabled',
                  enabled: false,
                  child: Text('Disabled'),
                ),
                const CupertinoMenuActionItem<String>(
                  icon: Icon(
                    CupertinoIcons.triangle,
                    semanticLabel: 'Triangle',
                  ),
                  value: 'Triangle',
                  child: Text('Triangle'),
                ),
                const CupertinoMenuActionItem<String>(
                  icon: Icon(
                    CupertinoIcons.square,
                    semanticLabel: 'Square',
                  ),
                  value: 'Square',
                  child: Text('Square'),
                ),
                const CupertinoMenuActionItem<String>(
                  icon: Icon(
                    CupertinoIcons.circle,
                    semanticLabel: 'Circle',
                  ),
                  value: 'Circle',
                  child: Text('Circle'),
                ),
                const CupertinoMenuActionItem<String>(
                  icon: Icon(
                    CupertinoIcons.star,
                    semanticLabel: 'Star',
                  ),
                  value: 'Star',
                  child: Text('Star'),
                ),
                const CupertinoMenuLargeDivider(),
                const CupertinoMenuItem<String>(
                  value: 'Delete',
                  isDestructiveAction: true,
                  trailing: Icon(
                    CupertinoIcons.delete,
                    semanticLabel: 'Delete',
                  ),
                  child: Text('Delete'),
                ),
              ];
            },
          ),
        ),
      ),
    );
  }
}

```

https://github.com/flutter/flutter/assets/59215665/ae65d6a8-7094-4352-877a-22e623693783

@MitchellGoodwin 
## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by MitchellGoodwin** [comment]
Thank you very much for putting this together! I'll block out some time to go through the design.

On basing it off of the pull_down_package, that might be fine, but the most foolproof way to make the lawyers happy is if @notDmDrl was willing to sign a CLA. If not, that's perfectly fine and we can see what guidance is available.

**Comment by notDmDrl** [comment]
@MitchellGoodwin @davidhicks980 I signed CLA

**Comment by MitchellGoodwin** [comment]
> @MitchellGoodwin @davidhicks980 I signed CLA

Thank you!

**Review by Piinks — COMMENTED** [review]
Welcome @davidhicks980

File: packages/flutter/lib/src/cupertino/menu.dart [hunk 1]
```diff
@@ -0,0 +1,4193 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// ignore_for_file: unused_element
```
Comment (Piinks) [inline packages/flutter/lib/src/cupertino/menu.dart:5]: Flyby comment from triage - this should be removed. :)

**Comment by Hixie** [comment]
@notDmDrl just for the complete avoidance of doubt, can you confirm whether you are ok with this code landing with the usual Flutter license and copyright rather than your MIT license and copyright?

File: packages/flutter/lib/src/cupertino/menu.dart (hunk 1 above)
Comment (davidhicks980) [reply packages/flutter/lib/src/cupertino/menu.dart]: Done!

**Comment by notDmDrl** [comment]
@Hixie, yes, I am ok with it :)

If there is anything that needs to be done with my package for this PR to land, please let me know

**Comment by davidhicks980** [comment]
Thanks @Piinks and team for taking a look and giving feedback! I'm still moving things around and fixing issues so I appreciate you all directing me towards any code I may have forgotten to prune. Also, some features that were missing have since been added -- in particular, sticky submenu headers. Whenever I have a moment, I'll draft up an overview of the mechanics of the menu in case any of my comments are ambiguous.

https://github.com/flutter/flutter/assets/59215665/2c1cefc4-5235-4a26-84d4-036eaa0584d3

**Review by Piinks — COMMENTED** [review]
As this is still a draft I have not been through it entirely, but I wanted to ask:

1. Why should this be in the framework instead of in a separate package? We typically ask this in so far as weighing the maintenance burden of adopting this into the framework.
2. Can it be split up into multiple PRs so it is easier to review? I think I saw mention of different features that have been added here. Landing a basic version and adding features in follow up PRs might be easier to do, since reviewers can better focus on one aspect at a time rather than everything all at once.

**Comment by davidhicks980** [comment]
> As this is still a draft I have not been through it entirely, but I wanted to ask:
> 
> 1. Why should this be in the framework instead of in a separate package? We typically ask this in so far as weighing the maintenance burden of adopting this into the framework.

Keeping it as a separate package is something I considered, and it could be done. Flutter is massive, and each additional component adds a maintenance and burden. That said, there are a few reasons that push me towards adding the menu to the framework:

1. Menus are used in most default iOS apps, so it does add feature parity to the Cupertino library. For example, the native filesystem menu on iOS uses nested layers for view options: image
2. The native menu is not something that can be easily replicated by a solo developer, and I couldn't find a good analog component within the Flutter codebase. One of the features I love about Flutter is the ease with which one can utilize the Flutter source code in their own apps. While components like the MenuAnchor offer a simple desktop-style dropdown menu, I think it could be helpful to demonstrate how to handle complex component interactions.
3. Some of the features in this PR could be applied to other Cupertino widgets. For example, the CupertinoPanListener can apply drag-hover effects found on iOS components to any class that mixes in PanTarget:
```dart
/// Can be mixed into a [State] to receive callbacks when a pointer enters or
/// leaves a [PanTarget]. The [PanTarget] is should be an ancestor of a
/// [CupertinoPanListener].
mixin PanTarget<T extends StatefulWidget> on State<T> {
  /// Called when a pointer enters the [PanTarget]. Return true if the pointer
  /// should be considered "on" the [PanTarget], and false otherwise (for
  /// example, when the [PanTarget] is disabled).
  bool didPanEnter();

  /// Called when the pointer leaves the [PanTarget]. If [pointerUp] is true,
  /// then the pointer was lifted while pressed on this menu item.
  void didPanLeave(bool pointerUp);
}
```

<video width="600" alt="video" src="https://github.com/flutter/flutter/assets/59215665/4493a615-983b-44a1-9812-a2c972ab0310" ></video>

4. I'm sure every open source contributor says this, but I'll keep maintaining this if/when it lands. Flutter is a really interesting project, so I have a lot of fun working with it.

With all of that said, I'm also aware of how much work maintaining a codebase as large as Flutter is, and how inconsistent community contributors of any open source project can be, so let me know what you are thinking. I don't want burden you all with more work.


> 2. Can it be split up into multiple PRs so it is easier to review? I think I saw mention of different features that have been added here. Landing a basic version and adding features in follow up PRs might be easier to do, since reviewers can better focus on one aspect at a time rather than everything all at once.

Of course -- would something like this work?
* PR 1 - CupertinoMenuButton, CupertinoMenu, CupertinoMenuItem, CupertinoMenuDivider, CupertinoMenuLargeDivider, CupertinoMenuController, showCupertinoMenu
* PR 2 - CupertinoMenuActionItem (horizontal menu items)
* PR 3 - CupertinoMenuTitle (small section title)
* PR 4 - CupertinoStickyMenuHeader
* PR 5 - CupertinoDragListener (drag highlight)
* PR 6 - CupertinoNestedMenu, CupertinoNestedMenuItemAnchor

I would submit each additional PR after the previous lands. I was actually considering this but was afraid to submit a partially-completed menu, so it's good to hear you had the same thought!

**Comment by davidhicks980** [comment]
Sounds great.

> Rather than committing unfinished portions, or parts that won't work until other changes land, why not land a basic version of the menu that works, then follow after it with additional features and stylings like dividers and headers? We typically do this with large features, so basic functionality is available, and then we add more bells and whistles. :)

Whoops, I meant this, but you said it a lot more clearly!

> I only skimmed this a couple of times so far since it is so large, but is this using MenuAnchor? It is meant to be a primitive for building menus such as these, and may simplify a lot of the code here (as well as reduce that burden of maintenance and bug fixes since it could re-use a bunch of logic 😉 ).

I'm asking this here but it could also be asked in the first PR. I spent some time trying to wrap this around the MenuAnchor, but there were a few problems. The implementation appears to have changed since I last took a look -- I think it's now using an OverlayPortal instead of an OverlayEntry, and the bulk of the widget has moved out of the _MenuAnchorState._open method. However, it'd still be hard to work around the following:

1. There was some material-specific code to work around. For example, _Submenu has padding and constraints that are determined based on Material-specific theming. This is not a huge problem, and probably could be worked around. Before, _Submenu was in the _open method, which made things more difficult.

```dart
@override
Widget build(BuildContext context) {
  Widget child = OverlayPortal(
    controller: _overlayController,
    overlayChildBuilder: (BuildContext context) {
     return _Submenu( // Submenu
        anchor: this,
        menuStyle: widget.style,
        alignmentOffset: widget.alignmentOffset ?? Offset.zero,
        menuPosition: _menuPosition,
        clipBehavior: widget.clipBehavior,
        menuChildren: widget.menuChildren,
        crossAxisUnconstrained: widget.crossAxisUnconstrained,
      );
    },
    child: _buildContents(context),
  );
//...
```

2. A bigger issue was with regard to the menu contents being wrapped in a SingleChildScrollView. From my understanding (which may be limited or out of date), sticky headers would be impossible without a CustomScrollView or a custom scrollable implementation. SingleChildScrollView would have been nice to use because it automatically becomes scrollable when the menu contents overflow. Currently, I'm measuring, which is certainly not a great way to determine! I just didn't want to add more code.

https://github.com/flutter/flutter/assets/59215665/69708f91-51cd-490a-bce6-63887c80df00

3. An even bigger issue is the jiu-jitsu to coordinate layer interactions using multiple overlays. iOS menu layers move in response to their children growing (see below). The CupertinoMenu component uses a FlowDelegate for almost all positioning and scaling, because the each layer could be painted per-frame without performing layout. Otherwise, I would have to either use a [MultiChildLayoutDelegate], which forces a layout per-frame (let me know if I'm wrong here), or I would have to coordinate movement across multiple Overlays (the MenuAnchor method). From my understanding, Overlays are basically Stacks, so using a Position widget across each overlay would involve a per-frame layout. A Transform widget could be used, but it would basically be a FlowDelegate with additional complexity. By using a FlowDelegate and measuring each layer when it is initially opened, all animations can be driven based on the final size of the each menu layer without having to continuously relayout each layer. That said, the topmost layer is still driven by a ConstraintsAnimation, so continuous layout for the top layer cannot be avoided. 

I actually began working on this menu using overlays, but there were some more issues I came across. When scaling nested layers, the untransformed position of the nested anchor couldn't be accessed via context.renderObject(), because the position would be affected by individual widget scaling and the scale of the entire menu. Scaling of underlying layers led to fragile alignment between the top nested anchor and the bottom nested anchor, which need to align perfectly after the menu has fully closed. MatrixUtils could be used to perform an inverse transform, but complexity grew to a point where the FlowDelegate appeared to be a much simpler solution. 

https://github.com/flutter/flutter/assets/59215665/a89fae57-8089-42ed-9395-27877de14c36

https://github.com/flutter/flutter/assets/59215665/3263ba92-a06a-4aba-87cb-7e70954a86c1

4. This has less to do with the MenuAnchor, but I figured I'd explain it since I'm explaining some of the other design decisions. The iOS version appears to use a physics simulation. To create the same effect, the CupertinoMenu implementation uses a bespoke AnimationController whose AnimationStatus can be manually set, because Simulation animations will not otherwise report forwards or reverse progress. Each animation is added into a single animation that controls the entire menu's position, which is passed into the FlowDelegate for positioning. I found it more intuitive to use the animation itself as a source of truth regarding an opened/closed/opening/closing state rather than synchronizing an animation and a controller, but this may violate separation of concerns. 

```dart
// Runs a simulation on the given animation controller, and returns whether or
// not the simulation has stopped running.
Future<bool> _runSimulation({
  required _AnimationControllerWithStatusOverride controller,
  required SpringSimulation simulation,
  required bool forward,
}) {
  final Completer<bool> simulationCompleter = Completer<bool>();

// Each layer has an animation controller that serves as a source-of-truth 
// regarding the opened/opening/closed/closing status of the menu layer.
  controller
    ..overrideStatus(
        forward
        ? AnimationStatus.forward
        : AnimationStatus.reverse,
    )
    ..animateWith(simulation)
        .whenCompleteOrCancel(() {
          if (!controller.isAnimating) {
            controller.overrideStatus(
                forward
                ? AnimationStatus.completed
                : AnimationStatus.dismissed,
            );
            simulationCompleter.complete(true);
          } else {
            simulationCompleter.complete(false);
          }
        });
  return simulationCompleter.future;
}

// Sums all current layer animations. A function is used instead of a
// CompoundAnimation to make it easier to add/remove animations.
void _updateAnimationValue() {
  double value = 0;
  for (
    final _AnimationControllerWithStatusOverride controller
    in _layerIdToAnimation.values
  ) {
    value += controller.value;
  }

  _nestingAnimation?.value = value;
}
```

https://github.com/flutter/flutter/assets/59215665/ddd51ff9-7d4e-4919-8741-61744223e805

https://github.com/flutter/flutter/assets/59215665/81a59d1c-a812-4dc2-bd27-2c49ff502e16

Anyways, sorry for the essay! I'll start splitting things up and sending PRs to get feedback. Thank you for taking the time to look through everything -- hopefully we can start landing features!

**Comment by mark8044** [comment]
@davidhicks980 These screenshots look epic.  Anxiously awaiting this!  Thanks so much for your work 👍 🥇 👍 🥇

**Comment by davidhicks980** [comment]
Quick update on this: I did a small rewrite to use a MenuAnchor that exposed an "overlayChildBuilder" on the MenuAnchor base class and it seems to work well at the moment. Also removed most of the hacky bits. I'll submit a PR soon.

```dart
@protected
  Widget overlayChildBuilder(BuildContext context) {
    return _Submenu(
      anchor: this,
      menuStyle: widget.style,
      alignmentOffset: widget.alignmentOffset ?? Offset.zero,
      menuPosition: _menuPosition,
      clipBehavior: widget.clipBehavior,
      menuChildren: widget.menuChildren,
      crossAxisUnconstrained: widget.crossAxisUnconstrained,
    );
  }
```

**Comment by JaidynBluesky** [comment]
Any update on this? Would love for some way to try this out in the interim @davidhicks980, this is just what I need for my app.

**Comment by davidhicks980** [comment]
@JaidynBluesky If you need all of the features right now, you should be able to copy the menu_item.dart and the menu.dart: https://github.com/flutter/flutter/blob/6bd90293300da86e31d9f0644c5e21a8e52c1a82/packages/flutter/lib/src/cupertino/menu.dart
and 
https://github.com/flutter/flutter/blob/6bd90293300da86e31d9f0644c5e21a8e52c1a82/packages/flutter/lib/src/cupertino/menu_item.dart . 

Just put them somewhere in your project, remove the local imports (see below), and import whichever items you want:

```dart
/// You'll want to import the Cupertino library instead of these files 
import 'colors.dart';
import 'constants.dart';
import 'icons.dart';)
```

Simple example that should work (I'm not on a device I can test at the moment, so sorry in advance for any errors): 

```dart
 CupertinoMenuButton<int>(
    itemBuilder: (BuildContext context) {
      return <CupertinoMenuEntry<int>>[
       const CupertinoMenuItem<int>(value: 1, child: Text('A')),
       const CupertinoMenuItem<int>(value: 2, child: Text('B')),
       const CupertinoMenuItem<int>(value: 3, child: Text('C')),
        CupertinoNestedMenu<int>(
          itemBuilder: (BuildContext context) {
            return [
                 const CupertinoMenuItem<int>(value: 1, child: Text('D')),
                 const CupertinoMenuItem<int>(value: 2, child: Text('E')),
                 const CupertinoMenuItem<int>(value: 3, child: Text('F')),
            ];
          },
          subtitle: const Text('Subtitle'),
          title: const TextSpan(text: 'Nested Menu'),
        ),
      ];
    },
    child: Icon(CupertinoIcons.plus, size: 24)
  );
```

If you only need a simple menu, the new version base on MenuAnchor will probably be closer to the final version of the menu, and has many more tests: 
https://github.com/flutter/flutter/pull/143712

**Comment by Piinks** [comment]
Hey @davidhicks980, we just came across this in triage today. Is https://github.com/flutter/flutter/pull/143712 meant to replace this PR? If so, should we close this?

**Comment by davidhicks980** [comment]
@Piinks - yes to both! I'll go ahead and close.

**Pull request closed without merging**